README.md
[npm-url]: https://npmjs.org/package/private-bower
[npm-image]: https://img.shields.io/npm/v/private-bower.svg
[downloads-image]: https://img.shields.io/npm/dm/private-bower.svg
[total-downloads-image]:
https://img.shields.io/npm/dt/private-bower.svg
[codeship-url]: https://codeship.com/projects/54990
[codeship-image]: https://img.shields.io/codeship/662b04e0-7427-0132-ff21-2aca0eeadc1e/master.svg

private-bower [![NPM version][npm-image]][npm-url] [![Downloads][downloads-image]][npm-url]  [![Total Downloads][total-downloads-image]][npm-url] [![Tests][codeship-image]][codeship-url]
============
<img src="https://raw.githubusercontent.com/Hacklone/private-bower/master/site/logo.png" alt="private-bower" width="250" height="250" />

The ultimate private bower server.

<a href="https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=XD9FKUJYSPP74" target="_blank">
  <img src="https://www.paypalobjects.com/en_US/i/btn/btn_donate_LG.gif" />
</a> 

Twitter: <a href="https://twitter.com/private_bower">@private_bower</a>, <a href="https://twitter.com/hashtag/private_bower?src=hash">#private_bower</a>

- [Features](#features)
- [Features to come](#features-to-come)
- [Installation](#installation)
- [Parameters](#parameters)
- [Config file](#config-file)
- [Usage](#usage)
  - [Web interface](#web-interface)
  - [Project](#project)
  - [List packages](#list-packages)
  - [Search packages](#search-packages)
  - [Register package](#register-package)
  - [Register packages](#register-packages)
  - [Remove package](#remove-package)
  - [Remove packages](#remove-packages)
  - [Restart server](#restart-server)
  - [Refresh caches](#refresh-caches)
  - [Authentication](#authentication)
  - [Log4js configuration examples](#log4js-configuration-examples)
- [License](#license)
- [Tips for usage](#tips-for-usage)
- [Contributing](#contributing)

#Features

*   Register private packages
*   Fallback to public packages
*   Cache public registry
*   Cache public git repositories
*   Cache public svn repositories
*   Web UI with package details
*   Web UI package management
*   Blacklist public packages
*   Whitelist public packages

#Installation

Install
> npm install -g private-bower

Run
> private-bower

Run with config file specified
> private-bower --config ./myBowerConfig.json

If there's no private package with requested package name the servers calls through to the public bower registry.

#Parameters

| name       | description                                    |
|------------|------------------------------------------------|
| --help     | print out help                                 |
| --config   | path to config file (Must be a valid json)     |

#Config file

Must be a valid JSON
```javascript
{
    "registryFile": "./bowerRepository.json",
    "timeout": 144000,
    "server": {
        "port": 5678,
        "hostName": null,
        "siteBaseUrl": null,
    },
    "public": {
        "disabled": false,
        "registry": "http://bower.herokuapp.com/packages",
        "registryFile": "./bowerRepositoryPublic.json",
        "whitelist": [],
        "blacklist": []
    },
    "authentication": {
        "enabled": false,
        "key": "password"
    },
    "repositoryCache": {
        "cachePrivate": false,
        "git": {
            "enabled": false,
            "cacheDirectory": "./gitRepoCache",
            "host": "localhost",
            "port": 6789,
            "protocol": "git",
            "publicAccessURL" : null,
            "refreshTimeout": 10
        },
        "svn": {
            "enabled": false,
            "cacheDirectory": "./svnRepoCache",
            "host": "localhost",
            "port": 7891,
            "protocol": "svn",
            "publicAccessURL" : null,
            "refreshTimeout": 10
        }
    },
    "proxySettings" : {
        "enabled": false,
        "host": "proxy",
        "username": "name",
        "password" : "pass",
        "port": 8080,
        "tunnel": false
    },
    "log4js" : {
        "enabled": false,
        "configPath" : "log4js.conf.json"
    }
}
```

| name                                       | description                                                                          | default                               |
|--------------------------------------------|--------------------------------------------------------------------------------------|---------------------------------------|
| server.port                                | Port on which the private bower server will listen                                   | 5678 (process.env.PORT if set)        |
| server.hostName                            | Host name on which the private bower server will listen                              | null (process.env.IP if set)          |
| server.siteBaseUrl                         | Load private bower server on a specific path, useful for using a reverse proxy       | null                                  |
| registryFile                               | File for persisting private packages (must be a valid json)                          | ./bowerRepository.json                |
| timeout                                    | server package timeout                                                               | 144 000                               |
| public.disabled                            | Disable fallback feature for public packages                                         | false                                 |
| public.registry                            | Public bower registry's url                                                          | http://bower.herokuapp.com/packages/  |
| public.registryFile                        | File for persisting public packages (must be a valid json)                           | ./bowerRepositoryPublic.json          |
| public.whitelist                           | Define public packages that are allowed to be installed                              | \[\]                                  |
| public.blacklist                           | Define public packages that are not allowed to be installed                          | \[\]                                  |
| authentication.enabled                     | Authentication enabled for registering packages                                      | false                                 |
| authentication.key                         | Authentication key (Auth-Key header)                                                 | password                              |
| repositoryCache.(svn, git).enabled         | Public repository caching enabled                                                    | false                                 |
| repositoryCache.cachePrivate               | Also cache privately registered packages                                             | false                                 |
| repositoryCache.(svn, git).host            | Server's host name for repository access                                             | localhost                             |
| repositoryCache.(svn, git).port            | Port to open repository server on                                                    | 7891, 6789                            |
| repositoryCache.(svn, git).protocol        | Protocol the mirrored repositories will use                                          | git, svn, https, http                 |
| repositoryCache.(svn, git).publicAccessURL | Public address to access repository cache (useful if repository is behind an apache) | null                                  |
| repositoryCache.(svn, git).cacheDirectory  | Directory where the public repository cache will save repositories                   | ./svnRepoCache, ./gitRepoCache        |
| repositoryCache.(svn, git).refreshTimeout  | Time to wai between repository cache refresh (minutes)                               | 10 minutes                            |
| repositoryCache.(svn, git).parameters.X    | Custom parameters for git-daemon and svnserve                                        | undefined                             |
| proxySettings.enabled                      | Enable the proxy, use the proxy to call the bower remote repo                        | false                                 |
| proxySettings.host                         | Proxy host                                                                           | proxy                                 |
| proxySettings.username                     | Proxy username                                                                       | name                                  |
| proxySettings.password                     | Proxy password                                                                       | pass                                  |
| proxySettings.port                         | Proxy port                                                                           | 8080                                  |
| proxySettings.tunnel                       | Use tunnel?                                                                          | false                                 |
| log4js.enabled                             | Use log4js ?                                                                         | false                                 |
| log4js.configPath                          | Log4js configuration file. See: log4js-node for configuration options                | none                                  |




#Usage

##Web interface
Convenient way for viewing your packages in a browser.  The web interface will only list your private packages, it will 
not list the public packages if you have a public registry enabled.  However, when searching for packages in bower, the 
public ones will show up just fine.
   
> http://localhost:5678/

##Project
Within your project, you will need to create a .bowerrc file containing the URL of your private bower server:
```json
{
  "registry": "http://yourPrivateBowerRepo:5678",
  "timeout": 300000
}
```

If you are using private bower with `server.siteBaseURL` option, you need to add the same path the registry url in your .bowerrc file:

Config
```json
{
  "server": {
    "port": 6789,
    "hostName": "yourPrivateBowerRepo",
    "setBaseURL": "/my-private-bower"
  }
}
```

.bowerrc
```json
{
  "registry": "http://yourPrivateBowerRepo:6789/my-private-bower",
  "timeout": 300000
}
```

##List packages
GET
> bower-server:5678/packages

##Search packages
> bower search \[packageName\]

##Register package
> bower register \[packageName\] \[gitRepo\]

or
POST
> bower-server:5678/packages

> { "name": "package-name", "url": "git://repoPath" }

or
POST
> bower-server:5678/packages/\<package name\>

> { "url": "git://repoPath" }

##Register packages
POST
> bower-server:5678/packages

> [ { "name": "package-name", "url": "git://repoPath" } ]

##Remove package
DELETE
> bower-server:5678/packages/\<package name\>

##Remove packages
DELETE
> bower-server:5678/packages

> ["package-name"]

##Restart server
POST
> bower-server:5678/restart

##Refresh caches
POST
> bower-server:5678/refresh

##Authentication

Authentication can be enabled for the following features:
*   Register package
*   Register packages
*   Remove package
*   Remove packages
*   Restart server

Add ```Auth-Key``` header to request.
> Auth-Key = password


##Log4js configuration
>There are two appenders set in the example configuration.
>You need to remove one of the two if you want to use it.
>fileDate appender will write the log to a file which will be rotated daily.
>Console will write the logging to the console in the log4js format.
>You need to set the replaceConsole to true if you want to write the logging to the log4j appenders.
>See github.com/nomiddlename/log4js-node for more information

#Tips for usage
##Server as a service
- [Installing on Ubuntu](https://github.com/Hacklone/private-bower/wiki/Installing%20on%20Ubuntu)
- [Install as a Windows service](https://github.com/Hacklone/private-bower/wiki/Install%20as%20a%20Windows%20service)

##Use behind proxy
> git config --global url."https://".insteadOf git://

##Calling the API
- do not forget to set the ```Content-Type``` header to ```application/json```

#Contributing
Please read the rules of contributing on the [contribution page](https://github.com/Hacklone/private-bower/blob/master/CONTRIBUTING.md).

#License
> The MIT License (MIT)

> Copyright (c) 2014 Hacklone
> https://github.com/Hacklone

> Permission is hereby granted, free of charge, to any person obtaining a copy
> of this software and associated documentation files (the "Software"), to deal
> in the Software without restriction, including without limitation the rights
> to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
> copies of the Software, and to permit persons to whom the Software is
> furnished to do so, subject to the following conditions:

> The above copyright notice and this permission notice shall be included in all
> copies or substantial portions of the Software.

> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
> IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
> FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
> AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
> LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
> OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
> SOFTWARE.


CHANGELOG.md
# Changelog

## 1.0.1 - 2015.04.18.

- Refactored whole framework with some automated tests
- Added management to website - create and delete packages ([#126](https://github.com/Hacklone/private-bower/issues/126))
- Created new design ([#125](https://github.com/Hacklone/private-bower/issues/125))
- Added option to whitelist / blacklist packages ([#115](https://github.com/Hacklone/private-bower/issues/115))
- Added api for manually refresh repository caches ([#117](https://github.com/Hacklone/private-bower/issues/117))
- fixed git-daemon crashes ([#112](https://github.com/Hacklone/private-bower/issues/112))
- fixed confusing return codes ([#118](https://github.com/Hacklone/private-bower/issues/118))

## 1.1.0 - 2015.06.29.

- Added bower.json package details to web ui ([#109](https://github.com/Hacklone/private-bower/issues/109))
- Make private-bower closed network friendly ([#144](https://github.com/Hacklone/private-bower/issues/144))
- bug fix ([#148](https://github.com/Hacklone/private-bower/issues/148))

## 1.1.1 - 2015.07.26.

- Added help text to package detail errors ([#149](https://github.com/Hacklone/private-bower/issues/149))
- Added process.env.PORT and process.env.IP support ([#141](https://github.com/Hacklone/private-bower/issues/141))
- Added tip for API usage in README.md ([#154](https://github.com/Hacklone/private-bower/issues/154))
- Registering a public package for the first time, hit count fixed ([#157](https://github.com/Hacklone/private-bower/issues/157))

## 1.1.2 - 2015.08.09.

- Added configurable protocol to mirrored packages ([#161](https://github.com/Hacklone/private-bower/issues/161))
- Fixed error flow when git cloning fails ([#156](https://github.com/Hacklone/private-bower/issues/156))

## 1.1.4 - 2015.08.29.

- enable version consultation ([#168](https://github.com/Hacklone/private-bower/issues/168))
- Added stdout to log message in case util.exec() fails. ([#164](https://github.com/Hacklone/private-bower/issues/164))
- fixed JSON parse errors ([#170](https://github.com/Hacklone/private-bower/issues/170))
- never used utils.process.env.PORT ([#159](https://github.com/Hacklone/private-bower/issues/159))

## 1.1.5 - 2015.10.10.

- change registry url ([#179](https://github.com/Hacklone/private-bower/issues/179))

## 1.1.6 - 2015.11.08.

- Added some documentation to clear up confusion regarding public registry. ([#184](https://github.com/Hacklone/private-bower/issues/184))
- Show the git url of a package (as a link) ([#182](https://github.com/Hacklone/private-bower/issues/182))

## 1.1.7 - 2015.12.02.

- private-bower sometimes fails to refresh certain cached repositories ([#171](https://github.com/Hacklone/private-bower/issues/171))
- Support unvalued options for git/svn daemons. ([#200](https://github.com/Hacklone/private-bower/issues/200))
- Avoid https error on the website  ([#191](https://github.com/Hacklone/private-bower/issues/191))
- Several fixes by @royrico


lib/server.js
var PackageStore = require('./package-store');
var PublicPackageStore = require('./public-package-store');
var RepoCacheHandler = require('./repo-cache-handler');

var argv = require('optimist').argv;
var fs = require('fs');
var path = require('path');
var spawn = require('child_process').spawn;
var log4js = require('log4js');
var bodyParser = require('body-parser');
var express = require('express');
var app = express();

var logger = require('./logger');

var _config, _configPath, _configDirectory;
var _repoCacheEnabled;
var _packageStore, _publicPackageStore, _repoCacheHandler;

module.exports.run = _init;
function _init() {
    if(argv.h || argv.help) {
        logger.logHelp();

        process.exit();
    }

    _loadConfig();

    _packageStore = new PackageStore({
        persistFilePath: _config.registryFile
    });

    if(!_config.disablePublic) {
        _publicPackageStore = new PublicPackageStore(_config);

        if(_config.repositoryCache) {
            initPublicRepoCache();
        }
    }

    _initService();

    function initPublicRepoCache() {
        var repoCacheOptions = {};

        if(_config.repositoryCache.svn && _config.repositoryCache.svn.enabled) {
            repoCacheOptions.svn = {
                repoCacheRoot: path.resolve(_config.repositoryCache.svn.cacheDirectory || path.join(_configDirectory, './svnRepoCache')),
                hostName: _config.repositoryCache.svn.host || 'localhost',
                port: _config.repositoryCache.svn.port || 7891,
                refreshTimeout: _config.repositoryCache.svn.refreshTimeout || 10,
                parameters: _config.repositoryCache.svn.parameters
            };
        }

        if(_config.repositoryCache.git && _config.repositoryCache.git.enabled) {
            repoCacheOptions.git = {
                repoCacheRoot: path.resolve(_config.repositoryCache.git.cacheDirectory || path.join(_configDirectory, './gitRepoCache')),
                hostName: _config.repositoryCache.git.host || 'localhost',
                publicAccessURL: _config.repositoryCache.git.publicAccessURL || null,
                port: _config.repositoryCache.git.port || 6789,
                refreshTimeout: _config.repositoryCache.git.refreshTimeout || 10,
                parameters: _config.repositoryCache.git.parameters
            };
        }

        if(repoCacheOptions.svn || repoCacheOptions.git) {
            _repoCacheEnabled = true;

            _repoCacheHandler = new RepoCacheHandler(repoCacheOptions);
        }
    }
}

function _loadConfig() {
    var defaultConfigPath = path.join(__dirname, '../bower.conf.json');
    _configPath = path.resolve(argv.config || defaultConfigPath);

    if(!fs.existsSync(_configPath)) {
        logger.error('config file not found at ' + _configPath);
    }

    _configDirectory = path.join(_configPath, '../');

    var json = fs.readFileSync(_configPath).toString();
    _config = JSON.parse(json);

    //defaults
    _config.port = _config.port || 5678;
    _config.registryFile = path.resolve(_config.registryFile || path.join(_configDirectory, './bowerRepository.json'));

    if(_config.registryFilePublic) {
        _config.registryFilePublic = path.resolve(_config.registryFilePublic);
    }

    _config.timeout = _config.timeout || 2 * 60 * 1200;

    if(_config.log4js && _config.log4js.enabled)  {
        log4js.configure(_config.log4js.configPath);
    }
}

function _initService() {
    app.use(bodyParser());
    app.use(express.static(path.join(__dirname, '../site')));

    app.post('/registerPackage', function(req, res) {
        if(!authenticate(req, res)) {
            return;
        }

        _packageStore.registerPackages([
            {
                name: req.body.name,
                repo: req.body.repo
            }
        ]);

        res.send('ok');
    });

    app.post('/packages', function(req, res) {
        if(!authenticate(req, res)) {
            return;
        }

        _packageStore.registerPackages([
            {
                name: req.body.name,
                repo: req.body.url
            }
        ]);

        res.send(201);
    });

    app.post('/registerPackages', function(req, res) {
        if(!authenticate(req, res)) {
            return;
        }

        _packageStore.registerPackages(req.body.packages);

        res.send('ok');
    });

    app.post('/removePackage', function(req, res) {
        if(!authenticate(req, res)) {
            return;
        }

        var packages = [ req.body.name ];

        _packageStore.removePackages(packages);

        if(_repoCacheHandler) {
            _removePackagesFromRepoCaches(packages);
        }

        res.send('ok');
    });

    app.post('/removePackages', function(req, res) {
        if(!authenticate(req, res)) {
            return;
        }

        var packages = req.body.packages;

        _packageStore.removePackages(packages);

        if(_repoCacheHandler) {
            _removePackagesFromRepoCaches(packages);
        }

        res.send('ok');
    });

    function _removePackagesFromRepoCaches(packages) {
        var repoCaches = _repoCacheHandler.getRepoAllCaches();

        repoCaches.forEach(function(repoCache) {
            packages.forEach(function(packageName) {
                repoCache.removeRepo(packageName);
            });
        });
    }

    app.get('/packages', function(req, res) {
        var packages = [];

        for(var packageName in _packageStore.packages) {
            if(_packageStore.packages.hasOwnProperty(packageName)) {
                var item = _packageStore.packages[packageName];

                packages.push({
                    name: item.name,
                    repo: item.repo,
                    hits: item.hits
                });
            }
        }

        res.send(packages);
    });

    app.get('/packages/search/:name', function(req, res) {
        var searchName = req.params.name;
        var packages = _packageStore.searchPackage(searchName);

        function filterPublicPackages(pkg){
            return _config.publicPackagesWhitelist.indexOf(pkg.name) > -1;
        }

        if(!_config.disablePublic) {
            Array.prototype.push.apply(packages, !_config.publicPackagesWhitelist
                ? _publicPackageStore.searchPackage(searchName)
                : _publicPackageStore.searchPackage(searchName).filter(filterPublicPackages));
        }

        res.send(packages);
    });

    //bower service
    app.get('/packages/:name', function(req, res) {
        var packageName = req.params.name;
        var privatePackage = _packageStore.getPackage(packageName);

        if(privatePackage) {
            handlePrivatePackage();
        }
        else if(!_config.disablePublic &&
            (!_config.publicPackagesWhitelist || _config.publicPackagesWhitelist.indexOf(packageName) > -1)) {
            handlePublicPackage();
        }
        else {
            res.status(404).send('Not found');
        }

        function handlePrivatePackage() {
            if(_config.repositoryCache && _config.repositoryCache.cachePrivate) {
                if(privatePackage.cachedRepo) {
                    res.send({
                        name: packageName,
                        url: privatePackage.cachedRepo,
                        hits: privatePackage.hits
                    });
                }
                else if(_repoCacheHandler) {
                    cachePrivateRepoAndSend();
                }
                else {
                    sendPrivatePackage();
                }
            }
            else {
                sendPrivatePackage();
            }

            function cachePrivateRepoAndSend() {
                var repoCache = _repoCacheHandler.getRepoCache(privatePackage.repo);

                repoCache.cacheRepo(packageName, privatePackage.repo)
                    .then(function(pack) {
                        privatePackage.cachedRepo = pack.repo;
                        _packageStore.persistPackages();

                        res.send({
                            name: packageName,
                            url: privatePackage.cachedRepo,
                            hits: privatePackage.hits
                        });
                    })
                    .fail(sendPrivatePackage);
            }

            function sendPrivatePackage() {
                res.send({
                    name: packageName,
                    url: privatePackage.repo,
                    hits: privatePackage.hits
                });
            }
        }

        function handlePublicPackage() {
            var publicPackage = _publicPackageStore.getPackage(packageName);
            if(publicPackage) {
                if(_repoCacheEnabled) {
                    cachePublicRepo();
                }
                else {
                    res.send(publicPackage);
                }
            } 
            else {
                res.status(404).send('Not found');
            }

            function cachePublicRepo() {
                var repoCache = _repoCacheHandler.getRepoCache(publicPackage.url);

                repoCache.cacheRepo(packageName, publicPackage.url)
                    .then(function(pack) {
                        var privatePackage = {
                            name: packageName,
                            repo: pack.repo,
                            hits: publicPackage.hits
                        };

                        _packageStore.registerPackages([ privatePackage ]);

                        res.send({
                            name: privatePackage.name,
                            url: privatePackage.repo,
                            hits: privatePackage.hits
                        });
                    })
                    .fail(function() {
                        res.send(publicPackage);
                    });
            }
        }


    });

    app.post('/restart', function(req, res) {
        if(!authenticate(req, res)) {
            return;
        }

        logger.log('Shutting down server for restart');

        res.send('ok');
        server.close();

        logger.log('Restarting private-bower with config set to ' + _configPath);

        _shutDown();

        _startDetachedChildProcess('private-bower', ['--config',  _configPath]);
    });

    var server = app.listen(_config.port, function() {
        logger.log('Bower server started on port ' + _config.port);
    });
    server.timeout = _config.timeout;

    process.on('SIGINT', _doShutDown);

    function authenticate(req, res) {
        if(!_config.authentication || !_config.authentication.enabled) {
            return true;
        }

        if(req.get('Auth-Key') === _config.authentication.key) {
            return true;
        }

        res.status(401);
        res.send('Unauthorized');

        return false;
    }
}

function _startDetachedChildProcess(command, args) {
    var child = spawn(command, args, {
        detached: true
    });

    child.stdout.pipe(process.stdout);
    child.stderr.pipe(process.stderr);

    child.unref();
}

function _doShutDown() {
    _shutDown();

    process.exit();
}

function _shutDown() {
    logger.log('Shutting down private-bower');

    if(_repoCacheHandler) {
        _repoCacheHandler.getRepoAllCaches().forEach(function(repoCache) {
            repoCache.shutDown();
        });
    }
}

process.on('uncaughtException', function(err) {
    logger.log('Exception message:' + (err.stack || err.message));

    _doShutDown();
});
